File: 2953-Count Complete Substrings.py
class Solution:
    def countCompleteSubstrings(self, word: str, k: int) -> int:
        N = len(word)
        
        # window length:k, 2*k, 3*k, ..., 26*k
        res = 0
        
        def cal(left, right):
            length = right-left+1
            res = 0
            for l in range(1, 27):
                if l*k > length:
                    break

                tmp = [0]*26
                currLen = l*k
                leftBound = left
                
                for i in range(left, right+1):
                    tmp[ord(word[i])-ord('a')] += 1
                    
                    if i-leftBound >= currLen:
                        tmp[ord(word[i-currLen])-ord('a')] -= 1
                        leftBound+=1
                    
                    if i-leftBound >= currLen-1:
                        if all(v == k or v == 0 for v in tmp):
                            res += 1
                    
            return res
        
        start = 0
        
        for i in range(N-1):
            if abs(ord(word[i])-ord(word[i+1])) > 2:
                res += cal(start, i)
                start = i+1

        res += cal(start, N-1)
        
        return res

File: C++/2953-Count Complete Substrings.cpp
class Solution {
public:
    int countCompleteSubstrings(string word, int k) {
        int N = word.size();
        int start = 0;
        int res = 0;
        
        auto cal = [&word, k](int start, int end) {
            int count = 0;
            for (int t=1; t<=26; t++){
                int window = t*k;
                if (window > end-start+1) break;
                vector<int> tmp(26, 0);
                for (int i=start; i<start+window; i++){
                    tmp[word[i]-'a']++;
                }
                int valid = 0;
                for (int i=0; i<26; i++){
                    if (tmp[i] == k){
                        valid++;
                    }
                }
                if (valid == t) count++;

                for (int i=start+window; i<=end; i++){
                    tmp[word[i]-'a']++;
                    if (tmp[word[i]-'a'] == k) valid++;
                    else if (tmp[word[i]-'a'] == k+1) valid--;
                    
                    tmp[word[i-window]-'a']--;
                    if (tmp[word[i-window]-'a'] == k) valid++;
                    else if (tmp[word[i-window]-'a'] == k-1) valid--;
                    
                    if (valid == t) count++;
                }
            }
            return count;
        };
        
        for (int i=1; i<N; i++){
            if (abs(word[i]-word[i-1]) > 2){
                res += cal(start, i-1);
                start = i;
            }
        }
        res += cal(start, N-1);
        
        return res;
    }
};